### PR TITLE
Extended Multifit functionality

### PR DESCRIPTION
**Pull request by bendavid**
Extends ecal multifit functionality with possibility for

1) Dynamic pedestals
2) Inflated pedestal uncertainty at level of python configuration
3) Better handling of noise covariance matrix for gain switch rechits
4) Ability to remove single sample from the fit to mitigate slew rate limitation

More information and background in
https://indico.cern.ch/event/600314/contributions/2441465/attachments/1398120/2132187/multifit-Jan18-2017.pdf

All of the new options are disabled by default, and therefore this pull request is expected to give 1:1 identical results for all standard workflows (including HLT).  There may be a small cpu and/or memory increase as the size on the stack of some of the matrix types have been increased to accommodate the new functionality.

This should not necessarily be merged yet, pending further tests and understanding of plans for 2017 data/mc, but want to at least run through the automated tests and verify 1:1 identity with present settings.

@emanueledimarco @amassiro @lgray @VinInn

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/EcalRecAlgos/interface/EcalUncalibRecHitMultiFitAlgo.h [hunk 1]
```diff
@@ -25,17 +25,27 @@ class EcalUncalibRecHitMultiFitAlgo
...
   EcalUncalibRecHitMultiFitAlgo();
   ~EcalUncalibRecHitMultiFitAlgo() { };
-  EcalUncalibratedRecHit makeRecHit(const EcalDataFrame& dataFrame, const EcalPedestals::Item * aped, const EcalMGPAGainRatio * aGain, const SampleMatrix &noisecor, const FullSampleVector &fullpulse, const FullSampleMatrix &fullpulsecov, const BXVector &activeBX);
+  EcalUncalibratedRecHit makeRecHit(const EcalDataFrame& dataFrame, const EcalPedestals::Item * aped, const EcalMGPAGainRatio * aGain, const SampleMatrixGainArray &noisecors, const FullSampleVector &fullpulse, const FullSampleMatrix &fullpulsecov, const BXVector &activeBX);
   void disableErrorCalculation() { _computeErrors = false; }
   void setDoPrefit(bool b) { _doPrefit = b; }
   void setPrefitMaxChiSq(double x) { _prefitMaxChiSq = x; }
+  void setDynamicPedestals(bool b=true) { _dynamicPedestals = b; }
+  void setMitigateBadSamples(bool b=true) { _mitigateBadSamples = b; }
+  void setSelectiveBadSampleCriteria(bool b=true) { _selectiveBadSampleCriteria = b; }
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/interface/EcalUncalibRecHitMultiFitAlgo.h]: please remove default values or introduce set and unset methods if no-argument calls are desired.

File: RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h [hunk 2]
```diff
@@ -1,16 +1,30 @@
+#ifndef EigenMatrixTypes_h
+#define EigenMatrixTypes_h
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h]: please use a more specific include guard ```  RecoLocalCalo_EcalRecAlgos_EigenMatrixTypes_h ```

Comment (VinInn) [inline RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h]: or move them to DataFormats/Math

File: RecoLocalCalo/EcalRecAlgos/src/EigenMatrixTypes.cc [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#include "RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h"
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/src/EigenMatrixTypes.cc]: is this file actually needed? EigenMatrixTypes.h contains only typedefs and constexpr.

File: RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc [hunk 4]
```diff
@@ -79,51 +79,82 @@ PulseChiSqSNNLS::PulseChiSqSNNLS() :
   _maxiterwarnings(true)
 {
   
-  Eigen::initParallel();
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: https://eigen.tuxfamily.org/dox/TopicMultiThreading.html

_In the case your own application is multithreaded, and multiple threads make calls to Eigen, then you have to initialize Eigen by calling the following routine before creating the threads_

Is removal justified?

File: RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo.cc [hunk 5]
```diff
@@ -20,70 +25,142 @@ EcalUncalibRecHitMultiFitAlgo::EcalUncalibRecHitMultiFitAlgo() :
...
+      if (!dynamicPedestal && _addPedestalUncertainty>0.) {
+        //add fully correlated component to noise covariance to inflate pedestal uncertainty
+        noisecov += _addPedestalUncertainty*_addPedestalUncertainty*SampleMatrix::Ones();
+      }
+    }
+    else {
+      noisecov = SampleMatrix::Zero();
+      for (unsigned int gainidx=0; gainidx<noisecors.size(); ++gainidx) {
+        SampleGainVector mask = gainidx*SampleGainVector::Ones();
+        SampleVector pedestal = (gainsNoise.array()==mask.array()).cast<SampleVector::value_type>();
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo.cc]: for my education: which operator/method is actually called here and where is its doc? A quick search for ```operator== ``` in Eigen pointed only to something that returns bool.
this one ``` (x==y)``` apparently has a non-trivial type.

File: RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc [hunk 6]
```diff
@@ -404,6 +420,28 @@ bool PulseChiSqSNNLS::NNLS() {
...
 
+void PulseChiSqSNNLS::NNLSUnconstrainParameter(Index idxp) {
+  
+  aTamat.col(_nP).swap(aTamat.col(idxp));
+  aTamat.row(_nP).swap(aTamat.row(idxp));
+  _pulsemat.col(_nP).swap(_pulsemat.col(idxp));
+  std::swap(aTbvec.coeffRef(_nP),aTbvec.coeffRef(idxp));
+  std::swap(_ampvec.coeffRef(_nP),_ampvec.coeffRef(idxp));
+  std::swap(_bxs.coeffRef(_nP),_bxs.coeffRef(idxp));
+  ++_nP;
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: there was also ```      wvec.tail(nActive) = updatework.tail(nActive);  ``` in what's replaced by NNLSUnconstrainParameter call.
Was that essential?

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc [hunk 7]
```diff
@@ -128,7 +135,14 @@ EcalUncalibRecHitWorkerMultiFit::set(const edm::EventSetup& es)
         // for the time correction methods
         es.get<EcalTimeBiasCorrectionsRcd>().get(timeCorrBias_);
 
-        int nnoise = noisecovariances->EBG12SamplesCorrelation.size();
+        int nnoise = SampleVector::RowsAtCompileTime;
+        SampleMatrix &noisecorEBg12 = noisecors_[1][0];
```
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc]: noisecorEBg12 etc were previously explicitly initialized to SampleMatrix::Zero().
I don't clearly see noisecors_ initialized here (and docs appear to say that the default constructed matrix is uninitialized)

File: RecoLocalCalo/EcalRecAlgos/interface/EcalUncalibRecHitMultiFitAlgo.h (hunk 1 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/interface/EcalUncalibRecHitMultiFitAlgo.h]: Ok will do that.

File: RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h (hunk 2 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/interface/EigenMatrixTypes.h]: Yes ok I'll leave them here and just change the ifndef as suggested.
The actual numbers which are here are pretty detector and algorithm specific, so I think it makes sense to keep them here anyways.

File: RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc (hunk 4 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: Yes, this can be removed as per the discussion at https://hypernews.cern.ch/HyperNews/CMS/get/swDevelopment/3290/1/1/1/1/1/1/1/1.html and the change here https://github.com/cms-externals/eigen/pull/1

@Dr15Jones can comment further

Comment (Dr15Jones) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: It appears @bendavid is correct.

File: RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo.cc (hunk 5 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/src/EcalUncalibRecHitMultiFitAlgo.cc]: It doesn't show up in the documentation because it's aliased in a strange way with macros inside the code, but it's pointing back to http://eigen.tuxfamily.org/dox/classEigen_1_1ArrayBase.html#a591a8024b2762dfdc4947cdcb9f74fdd

So it's an element wise comparison and returns a vector of bools (which is why the cast is needed)

File: RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc (hunk 6 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: The wvec vector was actually vestigial to a previous version of the implementation and is not needed at all (this pull request in fact removes it entirely https://github.com/cms-sw/cmssw/pull/17205/files#diff-08eda0e31e35729f5f2d2af294df5134L63)

File: RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc (hunk 7 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc]: Ok, yes this is maybe confusing.  The code here creates a non-const reference &noisecorEBg12 to a particular element of noisecors_, and likewise for the other elements, which are then all initialized in the existing code at
https://github.com/bendavid/cmssw/blob/f4837b2132cd150cdaf934fadf7e5ae07f3ff004/RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerMultiFit.cc#L146-L156

File: RecoLocalCalo/EcalRecAlgos/src/EigenMatrixTypes.cc (hunk 3 above)
Comment (bendavid) [inline RecoLocalCalo/EcalRecAlgos/src/EigenMatrixTypes.cc]: Yes, agreed that it is not needed, will remove it.

**Comment by bendavid** [comment]
Minor cleanup added according to comments.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc [hunk 8]
```diff
@@ -241,17 +272,18 @@ bool PulseChiSqSNNLS::Minimize(const SampleMatrix &samplecor, double pederr, con
...
 
-bool PulseChiSqSNNLS::updateCov(const SampleMatrix &samplecor, double pederr, const FullSampleMatrix &fullpulsecov) {
+bool PulseChiSqSNNLS::updateCov(const SampleMatrix &samplecov, const FullSampleMatrix &fullpulsecov) {
  
   const unsigned int nsample = SampleVector::RowsAtCompileTime;
   const unsigned int npulse = _bxs.rows();
 
-  const double pederr2 = pederr*pederr;
-  _invcov = pederr2*samplecor; //
+  _invcov = samplecov; //
```
Comment (slava77) [inline RecoLocalCalo/EcalRecAlgos/src/PulseChiSqSNNLS.cc]: it looks like this part and the way samplecov is now computed upstream lead to some differences at numerical precision level observed in DQM comparisons.

pederr2 was ``` (pederr*gainratio)*(pederr*gainratio) ```, while now it is ```(pederr* pederr* gainratio* gainratio)```

the other part of the logic that appears different is ``` abs(bx)>=100 ``` added in a few places, but IIUC, this never happens with the current defaults.
@bendavid 
please confirm

**Comment by slava77** [comment]
looking at jenkins DQM plots: there are changes starting from ECAL energies with differences at the level of float math precision

I ran 136.761 with 200 events locally and see
e.g. 
```
en->Scan("EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy():\
eo.EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy():\
EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy()-eo.EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy()",\
 "EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy()!=eo.EcalRecHitsSorted_ecalRecHit_EcalRecHitsEE_reRECO.obj.obj.energy()")
***********************************************************
*    Row   * Instance * EcalRecHi * eo.EcalRe * EcalRecHi *
***********************************************************
*       10 *      477 * 0.0106320 * 0.0106320 * 9.313e-10 *
*       11 *      464 * 2.8491241 * 2.8491237 * 4.768e-07 *
*       16 *      940 * 0.0003039 * 0.0003039 * -2.91e-11 *
*       17 *      353 * 0.0058631 * 0.0058631 * 4.656e-10 *
*       20 *      392 * 0.3206158 * 0.3206158 * -2.98e-08 *
*       21 *      537 * 6.178e-05 * 6.178e-05 * 7.275e-12 *
*       22 *       28 * 0.0018613 * 0.0018613 * 2.328e-10 *
```
there are 43 hits with different energies in EE and 123 in EB.
These differences propagate downstream: ~12 pfCandidates have different momentum.

If this is all just from pederr2 (noisecov) computed with a semantically different formula https://github.com/cms-sw/cmssw/pull/17205#discussion_r97231340
I can sign off on this.

If there is some other logic change that might create larger differences for some rare hits (e.g. the ones with gainswitch), then it would be better to understand what the effect on those would be.

**Comment by bendavid** [comment]
Agreed that changes at the level of numerical precision are possible/expected.  Also confirmed that abs(bx)>=100 never occurs presently (this only happens if dynamic pedestals are enabled)

The effect on gain switch rechits should be visible in an eg. 1tev electron/photon gun or very high mass Zprime->ee sample, but here also differences for standard workflows should be only at the level of numerical precision.  (To be checked whether there are already dqm plots which isolate the gain switch rechits, otherwise we probably need to explicitly check by hand, since there are some technical changes here which only affect the gain switch case, even if they should not change anything beyond numerical precision with present settings as said.)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17198	a4d9948
- code changes are in line with the description
- jenkins tests pass and comparisons with the baseline show differences in ECAL energies at the level of numerical precision (expected from the changes done) which are also visible in downstream objects
- local test with 1TeV ele gun shows only differences at numerical precision level

**Comment by davidlange6** [comment]
+1
